File: allegro_hand_teleop/scripts/allegro_hand_keys_to_joints.py
#!/usr/bin/env python
import copy
import rospy
from std_msgs.msg import String
from sensor_msgs.msg import JointState

# ---------------- CLASS ----------------
class AllegroHandKeyToJointsNode():
    def __init__(self):
        print('AllegroHandKeyToJointsNode: initializing node')

        self.ready = False
        self.joint_states = None

        # Initialize Variables
        self.joint_upper_limit = [
            0.57, 1.71, 1.809, 1.718,  # Index Finger
            0.57, 1.71, 1.809, 1.718,  # Middle Finger
            0.57, 1.71, 1.809, 1.718,  # Pinky Finger
            1.4968, 1.13, 1.633, 1.81991  # Thumb
        ]
        self.joint_lower_limit = [
            -0.57, -0.296, -0.274, -0.327,  # Index Finger
            -0.57, -0.296, -0.274, -0.327,  # Middle Finger
            -0.57, -0.296, -0.274, -0.327,  # Pinky Finger
            0.36357, -0.20504289, -0.2897, -0.2622  # Thumb
        ]
        self.joint_position = [
            0, 0, 0, 0,
            0, 0, 0, 0,
            0, 0, 0, 0,
            0, 0, 0, 0
        ]
        self.joint_velocity = [
            0, 0, 0, 0,
            0, 0, 0, 0,
            0, 0, 0, 0,
            0, 0, 0, 0
        ]
        self.joint_names = [
            'jif1', 'jif2', 'jif3', 'jif4',
            'jmf1', 'jmf2', 'jmf3', 'jmf4',
            'jpf1', 'jpf2', 'jpf3', 'jpf4',
            'jth1', 'jth2', 'jth3', 'jth4'
        ]
        self.finger = "thumb"
        
        self.desired_joint_state = JointState()
        self.desired_joint_state.name = self.joint_names

        # ROS Infrastructure
        topic = 'keys'
        self.sub_keys = rospy.Subscriber(topic, String, self.keys_callback)
        topic = 'joint_cmd'
        self.pub_joint_cmd = rospy.Publisher(topic, JointState, queue_size=10)
        topic = 'joint_states'
        self.sub_joint_states = rospy.Subscriber(topic, JointState, self.joint_states_callback)

        # Wait to get ready
        print('Waiting for joint_states...')
        while (not self.ready):
            pass
        print('Received first joint_states message')

        self.joint_position = list(self.joint_states.position)

        rospy.sleep(1)


    def keys_callback(self, msg):
        if not self.ready:
            return

        i = 0.05
        key = msg.data
        unlimited_joints = copy.copy(self.joint_position)
        if(key=='1'):
            self.finger = "thumb"
        elif(key=='2'):
            self.finger = "index"
        elif(key=='3'):
            self.finger = "middle"
        elif(key=='4'):
            self.finger = "pinky"
        elif(key=="z"):
            unlimited_joints = [
                0, 0, 0, 0,
                0, 0, 0, 0,
                0, 0, 0, 0,
                0, 0, 0, 0
            ]
        elif(key=="m"):
            unlimited_joints = [
                0.10, 1.60, 1.81, 1.72,
                4.45e-06, 1.21e-06, 3.30e-08, 2.62e-08,
                0.1, 1.40, 1.81, 1.72,
                1.46, 0.25, 0.89, 1.89
            ]
        elif(key=="x"):
            unlimited_joints = list(self.joint_states.position)

        if self.finger=="thumb":
            if(key=="q"):
                unlimited_joints[0+12] += i
            elif(key=="a"):
                unlimited_joints[0+12] -= i
            
            elif(key=="w"):
                unlimited_joints[1+12] += i
            elif(key=="s"):
                unlimited_joints[1+12] -= i
         
            elif(key=="e"):
                unlimited_joints[2+12] += i
            elif(key=="d"):
                unlimited_joints[2+12] -= i
         
            elif(key=="r"):
                unlimited_joints[3+12] += i
            elif(key=="f"):
                unlimited_joints[3+12] -= i
        elif self.finger=="index":
            if(key=="q"):
                unlimited_joints[0] += i
            elif(key=="a"):
                unlimited_joints[0] -= i
            
            elif(key=="w"):
                unlimited_joints[1] += i
            elif(key=="s"):
                unlimited_joints[1] -= i
         
            elif(key=="e"):
                unlimited_joints[2] += i
            elif(key=="d"):
                unlimited_joints[2] -= i
         
            elif(key=="r"):
                unlimited_joints[3] += i
            elif(key=="f"):
                unlimited_joints[3] -= i
        elif self.finger=="middle":
            if(key=="q"):
                unlimited_joints[0+4] += i
            elif(key=="a"):
                unlimited_joints[0+4] -= i
            
            elif(key=="w"):
                unlimited_joints[1+4] += i
            elif(key=="s"):
                unlimited_joints[1+4] -= i
         
            elif(key=="e"):
                unlimited_joints[2+4] += i
            elif(key=="d"):
                unlimited_joints[2+4] -= i
         
            elif(key=="r"):
                unlimited_joints[3+4] += i
            elif(key=="f"):
                unlimited_joints[3+4] -= i
        elif self.finger=="pinky":
            if(key=="q"):
                unlimited_joints[0+8] += i
            elif(key=="a"):
                unlimited_joints[0+8] -= i
            
            elif(key=="w"):
                unlimited_joints[1+8] += i
            elif(key=="s"):
                unlimited_joints[1+8] -= i
         
            elif(key=="e"):
                unlimited_joints[2+8] += i
            elif(key=="d"):
                unlimited_joints[2+8] -= i
         
            elif(key=="r"):
                unlimited_joints[3+8] += i
            elif(key=="f"):
                unlimited_joints[3+8] -= i

        # Limit position
        self.joint_position = self.limit_joints(unlimited_joints)

        # Publish desired JointState
        self.desired_joint_state.header.stamp = rospy.Time.now()
        self.desired_joint_state.position = self.joint_position
        self.pub_joint_cmd.publish(self.desired_joint_state)


    def joint_states_callback(self, msg):
        self.joint_states = msg
        self.ready = True


    def limit_joints(self, q):
        limited = copy.copy(q)
        for i in range(16):
            if q[i] > self.joint_upper_limit[i]:
                limited[i] = self.joint_upper_limit[i]
            elif q[i] < self.joint_lower_limit[i]:
                limited[i] = self.joint_lower_limit[i]
        return limited


if __name__ == '__main__':
    print('starting allegro_hand_keys_to_joints')
    rospy.init_node('allegro_hand_keys_to_joints', disable_signals=True)
    node = AllegroHandKeyToJointsNode()

    try:
        rospy.spin()
    except:  # rospy.ROSInterruptException:
        print('caught exception')
    print('exiting')

File: allegro_hand_teleop/src/allegro_hand_keys_to_joints.cpp
#include "allegro_hand_teleop/allegro_hand_keys_to_joints.h"

AllegroHandKeysToJointNode::AllegroHandKeysToJointNode(){
	// Initialize Variables
	ready = false;
	joint_state_cmd.name = joint_names;

	// ROS Infrastructure
	joint_cmd_pub = nh.advertise<sensor_msgs::JointState>(JOINT_CMD_TOPIC, 5);
	keys_sub = nh.subscribe(KEYS_TOPIC, 1, &AllegroHandKeysToJointNode::keysCallback, this);
	joint_state_sub = nh.subscribe(JOINT_STATE_TOPIC, 1, &AllegroHandKeysToJointNode::jointStateCallback, this);

	// Wait to get ready
	ROS_INFO("Waiting for joint_states...");
	while(!ready)
		ros::spinOnce();
	ROS_INFO("Received first joint_states message");

	// Initialize position
	joint_position = current_joint_state.position;
}

AllegroHandKeysToJointNode::~AllegroHandKeysToJointNode(){
	nh.shutdown();
}

void AllegroHandKeysToJointNode::keysCallback(const std_msgs::String::ConstPtr& msg){
	if (!ready)
		return;
	
	double i = 0.05;
	std::string key = msg->data.c_str();
	char key_char = (char) key[0];

	switch (key_char) {
	case '1':
		finger = FINGER_THUMB;
		break;
	case '2':
		finger = FINGER_INDEX;
		break;
	case '3':
		finger = FINGER_MIDDLE;
		break;
	case '4':
		finger = FINGER_PINKY;
		break;

	case 'z':
		joint_position = {
			0, 0, 0, 0,
			0, 0, 0, 0,
			0, 0, 0, 0,
			0, 0, 0, 0
		};
		break;
	case 'm':
		joint_position = {
			-0.07, 1.08, 0.60, 0.62, 
			-0.03, 1.11, 0.62, 0.70, 
			-0.07, 1.06, 0.69, 0.75, 
			 1.50, 0.32, 0.30, 0.86
		};
		break;
	case 'x':
		joint_position = current_joint_state.position;
		break;

	case 'q':
		if (finger == FINGER_THUMB)
			joint_position[0+12] += i;
		else if (finger == FINGER_INDEX)
			joint_position[0+0] += i;
		else if (finger == FINGER_MIDDLE)
			joint_position[0+4] += i;
		else
			joint_position[0+8] += i;
		break;
	case 'a':
		if (finger == FINGER_THUMB)
			joint_position[0+12] -= i;
		else if (finger == FINGER_INDEX)
			joint_position[0+0] -= i;
		else if (finger == FINGER_MIDDLE)
			joint_position[0+4] -= i;
		else
			joint_position[0+8] -= i;
		break;
	case 'w':
		if (finger == FINGER_THUMB)
			joint_position[1+12] += i;
		else if (finger == FINGER_INDEX)
			joint_position[1+0] += i;
		else if (finger == FINGER_MIDDLE)
			joint_position[1+4] += i;
		else
			joint_position[1+8] += i;
		break;
	case 's':
		if (finger == FINGER_THUMB)
			joint_position[1+12] -= i;
		else if (finger == FINGER_INDEX)
			joint_position[1+0] -= i;
		else if (finger == FINGER_MIDDLE)
			joint_position[1+4] -= i;
		else
			joint_position[1+8] -= i;
		break;
	case 'e':
		if (finger == FINGER_THUMB)
			joint_position[2+12] += i;
		else if (finger == FINGER_INDEX)
			joint_position[2+0] += i;
		else if (finger == FINGER_MIDDLE)
			joint_position[2+4] += i;
		else
			joint_position[2+8] += i;
		break;
	case 'd':
		if (finger == FINGER_THUMB)
			joint_position[2+12] -= i;
		else if (finger == FINGER_INDEX)
			joint_position[2+0] -= i;
		else if (finger == FINGER_MIDDLE)
			joint_position[2+4] -= i;
		else
			joint_position[2+8] -= i;
		break;
	case 'r':
		if (finger == FINGER_THUMB)
			joint_position[3+12] += i;
		else if (finger == FINGER_INDEX)
			joint_position[3+0] += i;
		else if (finger == FINGER_MIDDLE)
			joint_position[3+4] += i;
		else
			joint_position[3+8] += i;
		break;
	case 'f':
		if (finger == FINGER_THUMB)
			joint_position[3+12] -= i;
		else if (finger == FINGER_INDEX)
			joint_position[3+0] -= i;
		else if (finger == FINGER_MIDDLE)
			joint_position[3+4] -= i;
		else
			joint_position[3+8] -= i;
		break;
	}

	// Limit position
	limitJoints();

	// Send message
	joint_state_cmd.header.stamp = ros::Time::now();
	joint_state_cmd.position = joint_position;
	joint_cmd_pub.publish(joint_state_cmd);
}

void AllegroHandKeysToJointNode::jointStateCallback(const sensor_msgs::JointState& msg){
	current_joint_state = msg;
	ready = true;
}

void AllegroHandKeysToJointNode::limitJoints(){
	for (size_t i=0; i<ALLEGRO_DOF_JOINTS; i++){
		if (joint_position[i] > joint_upper_limit[i])
			joint_position[i] = joint_upper_limit[i];
		else if (joint_position[i] < joint_lower_limit[i])
			joint_position[i] = joint_lower_limit[i];
	}
}

int main(int argc, char** argv)
{
	ros::init(argc, argv, "allegro_hand_keys_to_joints");

	AllegroHandKeysToJointNode node;
	ros::spin();

	return 0;
}